pkg/agent/wireguard/client_linux.go

```diff
@@ -0,0 +1,378 @@
+		if err := client.ensureRouting(netlink.FAMILY_V4); err != nil {
+			return err
+		}
+	}
+	if client.enabledIPv6 {
+		if err := client.ensureRouting(netlink.FAMILY_V6); err != nil {
+			return err
+		}
+	}
+	wgdev, err := client.wgClient.Device(ifName)
+	if err != nil {
+		return err
+	}
+	client.privateKey = wgdev.PrivateKey
+	// WireGuard private key will be persistent when agent restarts. So antrea only needs to
+	// generate a new private key if it is empty (all zero).
+	if client.privateKey == zeroKey {
+		newPkey, err := wgtypes.GeneratePrivateKey()
+		if err != nil {
+			return err
+		}
+		client.privateKey = newPkey
+	}
+	cfg := wgtypes.Config{
+		PrivateKey:   &client.privateKey,
+		ListenPort:   &client.listenPort,
+		ReplacePeers: false,
+	}
+	patch, _ := json.Marshal(map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"annotations": map[string]string{
+				types.NodeWireGuardPublicKey: client.privateKey.PublicKey().String(),
+			},
+		},
+	})
+	if err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		_, err := client.k8sclient.CoreV1().Nodes().Patch(context.TODO(), client.nodeName, apitypes.MergePatchType, patch, metav1.PatchOptions{})
+		return err
+	}); err != nil {
+		return err
+	}
+
+	return client.wgClient.ConfigureDevice(ifName, cfg)
+}
+
+func (client *client) RemoveStalePeers(existingPeerPublickeys []string) error {
+	wgdev, err := client.wgClient.Device(ifName)
+	if err != nil {
+		return err
+	}
+	restoredPeerPublicKeys := make(map[wgtypes.Key]struct{})
+	for _, peer := range wgdev.Peers {
+		restoredPeerPublicKeys[peer.PublicKey] = struct{}{}
+	}
+	for _, e := range existingPeerPublickeys {
+		pubKey, err := wgtypes.ParseKey(e)
+		if err != nil {
+			klog.ErrorS(err, "Parse WireGuard public error")
+			continue
+		}
+		delete(restoredPeerPublicKeys, pubKey)
+	}
+	for k := range restoredPeerPublicKeys {
+		if err := client.deletePeerByPublicKey(k); err != nil {
+			klog.ErrorS(err, "Delete WireGuard peer error")
+		}
+	}
+	return nil
+}
+
+func (client *client) UpdatePeer(nodeName, publicKeyString string, nodeIPv4, nodeIPv6 net.IP) error {
+	pubKey, err := wgtypes.ParseKey(publicKeyString)
+	if err != nil {
+		return err
+	}
+	var endpoint string
+	// We can only set one endpoint for each peer. Shoud we prioritize using IPv6?
+	if client.enabledIPv4 && nodeIPv4 != nil {
+		endpoint = net.JoinHostPort(nodeIPv4.String(), strconv.Itoa(listenPort))
+	} else if client.enabledIPv6 && nodeIPv6 != nil {
+		endpoint = net.JoinHostPort(nodeIPv6.String(), strconv.Itoa(listenPort))
+	} else {
+		return fmt.Errorf("node IP missing: %s", nodeName)
+	}
+	if p, exist := client.peerByNodeName.Load(nodeName); exist {
+		peer := p.(wgtypes.PeerConfig)
+		if peer.PublicKey.String() != publicKeyString {
+			klog.InfoS("WireGuard peer public key changed", "nodeName", nodeName, "publicKey", publicKeyString)
+			// delete old peer by public key
+			if err := client.deletePeerByPublicKey(peer.PublicKey); err != nil {
```

Comment:
It seems you only need to cache the previous publicKey, instead of the entire peerConfig?